crates/bevy_ecs/src/component.rs

```diff
@@ -283,22 +284,48 @@ impl ComponentInfo {
 /// one `World` to access the metadata of a `Component` in a different `World` is undefined behavior
 /// and must not be attempted.
 #[derive(Debug, Copy, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
-pub struct ComponentId(usize);
+pub struct ComponentId {
+    index: u32,
+    target: u32,
+}
 
 impl ComponentId {
+    const NO_TARGET: u32 = u32::MAX;
+
     /// Creates a new [`ComponentId`].
     ///
     /// The `index` is a unique value associated with each type of component in a given world.
     /// Usually, this value is taken from a counter incremented for each type of component registered with the world.
     #[inline]
     pub const fn new(index: usize) -> ComponentId {
-        ComponentId(index)
+        ComponentId {
+            index: index as u32,
+            target: Self::NO_TARGET,
+        }
+    }
+
+    /// Creates a new relation [`ComponentId`].
+    ///
+    /// The relation is of the input `ComponentId` and the target is the given entity.
+    ///
+    /// ## Examples
+    /// ```rust,ignore
+    /// let child_of: ComponentId = todo!(); // Get the component id of `ChildOf`
+    /// let parent: Entity = todo!(); // Get some parent
+    /// let child_of_parent: ComponentId = child_of.relation(parent);
+    /// ```
+    #[inline]
+    pub const fn relation(self, target: Entity) -> Self {
+        ComponentId {
+            index: self.index,
+            target: target.index(),
```

Comment:
I don't think this is valid. Won't this lead to collisions when we have two relations targeting different entities with the same index but different generations?

Reply:
It would, and currently it will lead to noticeable collisions. The solution that Flecs uses (because it uses a very similar encoding) is that it applies cleanup policies so that when an entity is despawned, all relations to it are also removed, or the entities they're on are deleted, or the whole app panics (which is mainly used for components afaik). We're going to need either cleanup policies or some other mechanism to solve this.

Reply:
For the current PR it feels like we may just want to store an entire `Entity` in there to avoid shipping subtle bugs (knowing that we have a plan to clean up performance later).

Reply:
Switched to full entity and noted in the PR description.